MerkleMountainRange/src/merklemountainrange.rs

```diff
@@ -0,0 +1,322 @@
+// We do this as we require the peer to prove the hash of the node. Taking the above example, let's prune leave 1.
+//
+//                 /\
+//                /  \
+//               /    \
+//              /      \
+//             /        \
+//            /          \
+//           /            \
+//          /\             \
+//         /\ \            /\
+//        /  \ \          /  \
+//       /\   \ \        /\   \
+//      /  \   \ \      /  \   \
+//     /\  /\  /\ \    /\  /\  /\
+//    /\/\/\/\/\/\/\  /\/\/\/\/\/\
+//
+// Node 1 has now only been marked as pruned but we cannot remove it as of yet because we still require it to
+// prove node 2. When we prune node 2, the MMR looks as follows
+//
+//                 /\
+//                /  \
+//               /    \
+//              /      \
+//             /        \
+//            /          \
+//           /            \
+//          /\             \
+//         /\ \            /\
+//        /  \ \          /  \
+//       /\   \ \        /\   \
+//      /  \   \ \      /  \   \
+//     /\  /\  /\ \    /\  /\  /\
+//      /\/\/\/\/\/\  /\/\/\/\/\/\
+//
+// Although we have not removed node 1 and node 2 from the MMR, we cannot yet remove node 3 as we require node 3 for
+// the proof of node 6. Let's prune 4 and 5.
+//
+//                 /\
+//                /  \
+//               /    \
+//              /      \
+//             /        \
+//            /          \
+//           /            \
+//          /\             \
+//         /\ \            /\
+//        /  \ \          /  \
+//       /\   \ \        /\   \
+//      /  \   \ \      /  \   \
+//         /\  /\ \    /\  /\  /\
+//        /\/\/\/\/\  /\/\/\/\/\/\
+//
+// Now we removed 3 from the MMR
+
+use crate::merklenode::{Hashable, MerkleNode, ObjectHash};
+use std::collections::HashMap;
+
+pub struct MerkleMountainRange<T>
+where T: Hashable
+{
+    // todo convert these to a bitmap
+    mmr: Vec<MerkleNode>,
+    data: HashMap<ObjectHash, T>,
+}
+
+impl<T> MerkleMountainRange<T>
+where T: Hashable
+{
+    /// This function creates a new empty Merkle Mountain Range
+    pub fn new() -> MerkleMountainRange<T> {
+        MerkleMountainRange { mmr: Vec::new(), data: HashMap::new() }
+    }
+
+    /// This function returns a reference to the data stored in the mmr
+    /// It will return none if the hash does not exist
+    pub fn get_object(&self, hash: ObjectHash) -> Option<&T> {
+        self.data.get(&hash)
+    }
+
+    /// This function returns a mut reference to the data stored in the mmr
+    /// It will return none if the hash does not exist
+    pub fn get_mut_object(&mut self, hash: ObjectHash) -> Option<&mut T> {
+        self.data.get_mut(&hash)
+    }
+
+    /// This function returns the hash proof tree of a given hash.
+    /// If the given hash is not in the tree, the vec will be empty.
+    /// The Vec will be created in form of the child-child-parent
```

Comment:
It would be helpful to expand on what the hash proof is; This comment suggests that the proof is always 3 elements long, and that doesn't square with what I understand a hash proof to be.

Reply:
I expanded it a bit to try and stop that confusion. I however feel that this is not the place to explain what a hash proof is. I plan on adding validation of hashproofs as well.